crates/node/src/shims/abcipp_shim.rs

```diff
@@ -114,43 +113,89 @@ impl AbcippShim {
                 Req::EndBlock(_) => {
                     let begin_block_request =
                         self.begin_block_request.take().unwrap();
-                    let block_time = begin_block_request
-                        .header
-                        .time
-                        .try_into()
-                        .expect("valid RFC3339 block time");
 
-                    let tm_raw_hash_string = tm_raw_hash_to_string(
-                        begin_block_request.header.proposer_address,
-                    );
-                    let block_proposer = find_validator_by_raw_hash(
-                        &self.service.state,
-                        tm_raw_hash_string,
-                    )
-                    .unwrap()
-                    .expect(
-                        "Unable to find native validator address of block \
-                         proposer from tendermint raw hash",
-                    );
+                    let (process_proposal_result, processing_results) =
+                        match namada::core::hash::Hash::try_from(
+                            begin_block_request.hash,
+                        ) {
+                            Ok(block_hash) => {
+                                match self
+                                    .service
+                                    .state
+                                    .in_mem()
+                                    .process_proposal_cache
+                                    .get(&block_hash)
+                                {
+                                    Some((process_resp, res)) => (
+                                        // We already have the result of
+                                        // process proposal for this block
+                                        // cached in memory
+                                        process_resp.to_owned(),
+                                        res.iter()
+                                            .map(|res| res.to_owned().into())
+                                            .collect(),
+                                    ),
+                                    None => {
+                                        // Need to run process proposal to
+                                        // extract the data we need
+                                        // for finalize block (tx results)
+                                        let process_req = super::abcipp_shim_types::shim::request::CheckProcessProposal::from(begin_block_request.clone()).cast_to_tendermint_req(self.delivered_txs.clone());
```

Comment:
can we trry to fix formatting here? it's all fucked

Reply:
Done in 33a4049e0b0b87b6fd68b560d3d2e3a43663ea8e